doc/commands.md

```diff
@@ -128,48 +128,49 @@ Rewriting rules for definable symbols are declared using the `rule` command.
 ```
 rule add zero      $n ↪ $n
 rule add (succ $n) $m ↪ succ (add $n $m)
+rule mul zero      _  ↪ zero
 ```
 
-Note that rewriting rules can also be defined simultaneously, using the `with`
-keyword instead of the `rule` keyword for all but the first rule.
+Terms prefixed by the sigil `$` and `_` are pattern variables.
+
+**Higher-order pattern-matching**.
+Lambdapi allows higher-order pattern-matching following Miller’s patterns.
 
 ```
-rule add zero      $n ↪ $n
-with add (succ $n) $m ↪ succ (add $n $m)
+rule diff (λx, sin $F[x]) ↪ λx, diff (λx, $F[x]) x × cos $F[x]
 ```
 
-Pattern variables need to be prefixed by `$`.
-
-**Higher-order pattern-matching**.
-Lambdapi allows higher-order pattern-matching (not in full generality
-though):
+Patterns can contain abstractions `λx, _` and the user may
+attach an environment made of *distinct* bound variables to a pattern variable.
+The environment is a semicolon-separated list of variables enclosed in square
+brackets `[x;y;...]`.
+A term containing a bound variable `x` only matches pattern variables whose
+environment contain `x`, such as `$F[x]`.
```

Comment:
What do you think of: "Patterns can contain abstractions `λx, _` and the user may
attach an environment made of *distinct* bound variables to a pattern variable, to indicate on what bound variables it is allowed to depend.
The environment is a semicolon-separated list of variables enclosed in square
brackets `[x;y;...]`.
For instance, a term of the form `λx y,t` matches the pattern `λx y,$F[x]` only if `y` does not freely occur in `t`." ?

Reply:
I feel that the `it` in
> to indicate on what bound variables it is allowed to depend

is ambiguous. What about, then

> Patterns can contain abstractions `λx, _` and the user may attach an environment made of distinct bound variables to a pattern variable to indicate which bound variable can occur in the matched term. The environment is a semicolon-separated list of variables enclosed in square brackets `[x;y;...]`. For instance, a term of the form `λx y,t` matches the pattern `λx y,$F[x]` only if y does not freely occur in `t`.

and it seems to me that "freely" in the last line is superfluous; or there is a nuance I don't grasp.

Reply:
Well, strictly speaking, y can occur as bound variable...

Reply:
Ok for "which bound variable can occur in the matched term".

Reply:
Ah ok, introduced by another binder, I see I see, fine.